### PR TITLE
Convert replay CLI to typescript

### PR DESCRIPTION
**Pull request by ryanjduffy**

### CODE REVIEW

File: packages/replay/metadata/index.ts [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
 // Each known metadata block should have a sanitizer that will check the contents before the upload
 const handlers = {
   test: test.validate
 };
 
+type AllowedKey = keyof typeof handlers;
 const ALLOWED_KEYS = Object.keys(handlers);
 
+function isAllowedKey(key: string): key is AllowedKey {
+  if (ALLOWED_KEYS.includes(key)) {
```
Comment (markerikson) [inline packages/replay/metadata/index.ts:14]: :hand: tiny nitpick: could just be `return ALLOWED_KEYS.includes(key)` :)

File: packages/replay/metadata/index.ts [hunk 2]
```diff
@@ -24,9 +34,9 @@ function sanitize(metadata, opts = {}) {
     if (value === null || key.startsWith("x-")) {
       // passthrough null or userspace types
       updated[key] = value;
-    } else if (ALLOWED_KEYS.includes(key)) {
+    } else if (isAllowedKey(key)) {
       // validate known types
-      Object.assign(updated, handlers[key](metadata));
+      Object.assign(updated, handlers[key](metadata as any));
```
Comment (markerikson) [inline packages/replay/metadata/index.ts:39]: :speech_balloon: Probably a better way to do it here, but this is not safety-critical code, so YOLO

File: packages/replay/metadata/test.ts [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import { Struct } from "superstruct";
```
Comment (markerikson) [inline packages/replay/metadata/test.ts:1]: :hand: can probably use `import type { Struct }` here. Not critical, but nice to delineate what's a type vs what's a value

File: packages/replay/metadata/test.ts [hunk 4]
```diff
@@ -13,30 +15,30 @@ const isUuid = require("is-uuid");
...
   if (!metadata || !metadata.test) {
     throw new Error('Test metadata does not exist');
   }
 
   return init(metadata.test);
 }
 
-function init(data) {
-  const version = data.version || VERSION;
+function init(data: Record<string, unknown>) {
```
Comment (markerikson) [inline packages/replay/metadata/test.ts:40]: :hand: given that we have `Record<string, unknown>` popping up quite a bit, I'll often extract that into its own type just so I don't repeat myself

File: packages/replay/src/upload.ts [hunk 5]
```diff
@@ -1,20 +1,26 @@
...
+) {
   if (!gClient) {
     let { resolve } = gClientReady;
     gClient = new ProtocolClient(
       server,
       {
         async onOpen() {
           try {
-            await gClient.setAccessToken(accessToken);
+            await gClient?.setAccessToken(accessToken);
```
Comment (markerikson) [inline packages/replay/src/upload.ts:23]: :hand: We should be pretty sure it exists on this line, right? Looks like `!` would be more appropriate here

**Review by markerikson — APPROVED** [review]
Tiny nitpicks, but LGTM!

File: packages/replay/metadata/index.ts (hunk 2 above)
Comment (ryanjduffy) [reply packages/replay/metadata/index.ts:39]: Agreed. I spent a little time trying to figure out the right type but couldn't and I bailed on it.

**Comment by ryanjduffy** [comment]
Fixed up a few things based on @markerikson 's comments in c1b2f60 and then realized there were some ts issues when you bootstrapped lerna so I also added 3cee941 to get everything working so you can:

* `npm run bootstrap` to install all deps and cross-link packages
* `npm run typecheck` to run `tsc` for each bundle`
* and run `npm link` in the `dist/` folder of each package to use the local copy for testing

😓

**Pull request merged**